### ISSUE TITLE
Feature Request: Some way to generate identity matrices

### ISSUE THREAD

**Issue by litherum** [gpuweb/gpuweb#4026]
Identity matrices are super useful. We should have a builtin way to generate them.

**Comment by jeancolasp** [comment]
Additionally would love to see the inverse() function proposed in WGSL. Atm we do have to carry them though a Uniform buffer. Which is not big deal, but would find it convenient to just calculate the inverse of a matrix from within the shader directly. Inverse() being supported by other shading APIs.

**Comment by greggman** [comment]
> Atm we do have to carry them through a Uniform buffer.

That's not your only option. You can include the code to do an inverse in your shader. Plenty of examples around the net of doing it in other shading languages. [1](https://github.com/Unity-Technologies/Graphics/blob/b2792beb10a3d3f1529e0327c98620db8c9e98da/Packages/com.unity.render-pipelines.core/ShaderLibrary/Common.hlsl#L907) [2](https://github.com/Unity-Technologies/Graphics/blob/master/Packages/com.unity.render-pipelines.core/ShaderLibrary/Common.hlsl#L907) [3](https://gist.github.com/greggman/4772468da1121406abaa79c8514b23ef)

> supported by other shading APIs.

Checking,

| HLSL | GLSL 4.6 | GLSL ES 3.0 | MSL | Cg | SPIR-V |
| ---- | -------- | ----------- | --- | -- | ------ |
|  ⚪️  |     ⚪️    |             |     | ⚪️ |        |

Not 100% sure of those.

I'm not against the idea of adding inverse, just pointing out that AFAIK, all it's doing is inserting the code you would have inserted yourself so while I agree it would nice, you have a solution today.

**Comment by jeancolasp** [comment]
Yes, I agree but then the reason could go with small functions such as mix for instance. But inverse() is very common, and the less work we can do that's standard in 3D, the less error-prone the code is. Furthermore, I was surprised that transpose was in the spec but not inverse. I realize you need to draw a line between the built-in functions and what is left for the end programmer to him/herself, but if transpose is there, I'd expect to see an inverse().

The thing that I'd be more worried about the idea of calling an inverse in a shader is whether this is really a useful way of using your computing resources, considering that it only needs to be calculated before the frames get rendered really. That's a detail probably and I am not sure what optimization is done at the compiling/shader stage to ensure that the inverse is not calculated several times.

**Comment by ben-clayton** [comment]
> The thing that I'd be more worried about the idea of calling an inverse in a shader is whether this is really a useful way of using your computing resources, considering that it only needs to be calculated before the frames get rendered really (in the case of projection/view matrices at least). That's a detail, probably, and I am not sure what optimization could be done at the compiling/shader stage to ensure that the inverse would not be calculated several times (if not varying).

+1 to this. [Inverse is a **very expensive** operation](https://source.chromium.org/chromium/chromium/src/+/refs/heads/main:third_party/dawn/src/tint/reader/spirv/function.cc;l=4179-4269;drc=e0e0d24aaa54727dc0a8bc4b159ccdf80d3f5d8d), and I fear that users seeing a standard language builtin for this might assume there's some form of special hardware acceleration for this operation - which as far as I'm aware, there isn't.

The only counter argument I can think of to this is that you may have a compute shader that intends to offload matrix inversion from the CPU to the GPU. In this situation, implementing the inverse function in user-code is still a viable option - with exactly the same performance traits.

**Comment by kdashg** [comment]
Inverse is wildly different from transpose, from my point of view.
More importantly, inverse is very different from ergonomically generating identity matrices, and **if you want to propose an inverse builtin, I must insist you file a new issue.** :)

**Comment by jeancolasp** [comment]
Totally. These are different requests. But thought it could be a good way to start to open the discussion (even though it mixes the issues).

I will open an issue then regarding inverse(). At least it will be in the backlog if people want to discuss it (and I would agree this is not an important feature and even maybe one that shouldn’t be in a real-time graphics API.

**Comment by litherum** [comment]
> I will open an issue then regarding inverse()

I searched but I didn’t find one. Do you think you could link to it here when you open it?

**Comment by jeancolasp** [comment]
https://github.com/gpuweb/gpuweb/issues/4115 <https://github.com/gpuweb/gpuweb/issues/4115>

Sorry I wasn’t logged in as the same user so it’s not under my name.

**Comment by litherum** [comment]
Given the existing constructors, I think it makes sense for these to be a family of standalone functions. So, I think the only open question is how to spell them.

Recall that the way to spell the matrix type names is like `mat4x4<f32>`.

- Option 1: `identity4x4<f32>()`
- Option 2: `identity<4, 4, f32>()`
- Option 3: `identity4x4f32()` (a bit unfortunate given the author may have `typedef`ed `f32` to their own name, but they can't use their own name here)
- Option 4: `diagonal4x4(1.0)` (broadcasts the input to the diagonal, infers the element type from the input) (would be even better if we had default arguments so the author didn't actually have to write the literal `1.0`

(Also, keep in mind that we might want to add an `identity(T) -> T` function in the future, which uses the same name "identity" but in a different way.)

**Comment by ben-clayton** [comment]
> Option 3: `identity4x4f32()`

Given the existing [single letter suffixes for type aliases](https://www.w3.org/TR/WGSL/#matrix-types), wouldn't it me more consistent to be `identity4x4f()` ?

I'll throw another (likely controversial) proposal into the mix:

* Option 5: `mat4x4f.identity` / `mat4x4<f32>.identity`

**Comment by kdashg** [comment]
<details><summary><a href="https://github.com/gpuweb/gpuweb/wiki/WGSL-2023%E2%80%9012%E2%80%9005-Minutes#matrix-identity-httpsgithubcomgpuwebgpuwebissues4026-">WGSL 2023-12-05 Minutes</a></summary>

* KG: **M2** sounds right.

</details>
